// astro/dist/core/dev/index.js
import { polyfill } from "@astropub/webapi";
import { performance } from "perf_hooks";
import { createVite } from "../create-vite.js";
import { defaultLogOptions, info } from "../logger.js";
import * as vite from "vite";
import * as msg from "../messages.js";
async function dev(config, options = { logging: defaultLogOptions }) {
  var _a;
  const devStart = performance.now();
  polyfill(globalThis, {
    exclude: "window document"
  });
  const viteUserConfig = vite.mergeConfig({
    mode: "development",
    server: {
      host: config.devOptions.hostname
    }
  }, config.vite || {});
  const viteConfig = await createVite(viteUserConfig, { astroConfig: config, logging: options.logging, mode: "dev" });
  const viteServer = await vite.createServer(viteConfig);
  await viteServer.listen(config.devOptions.port);
  const address = viteServer.httpServer.address();
  const site = config.buildOptions.site ? new URL(config.buildOptions.site) : void 0;
  info(options.logging, "astro", msg.devStart({ startupTime: performance.now() - devStart }));
  info(options.logging, "astro", msg.devHost({ address, site, https: !!((_a = viteUserConfig.server) == null ? void 0 : _a.https) }));
  return {
    address,
    stop: () => viteServer.close()
  };
}
export {
  dev as default
};
