// astro/dist/core/render/result.js
var __defProp = Object.defineProperty;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
import { bold } from "kleur/colors";
import { canonicalURL as getCanonicalURL } from "../util.js";
import { isCSSRequest } from "./dev/css.js";
import { isScriptRequest } from "./script.js";
import { renderSlot } from "../../runtime/server/index.js";
import { warn } from "../logger.js";
function createResult(args) {
  var _a, _b;
  const { experimentalStaticBuild, origin, markdownRender, params, pathname, renderers, resolve, site: buildOptionsSite } = args;
  const result = {
    styles: new Set(),
    scripts: (_a = args.scripts) != null ? _a : new Set(),
    links: (_b = args.links) != null ? _b : new Set(),
    createAstro(astroGlobal, props, slots) {
      const site = new URL(origin);
      const url = new URL("." + pathname, site);
      const canonicalURL = getCanonicalURL("." + pathname, buildOptionsSite || origin);
      return {
        __proto__: astroGlobal,
        props,
        request: {
          canonicalURL,
          params,
          url
        },
        resolve(path) {
          if (experimentalStaticBuild) {
            let extra = `This can be replaced with a dynamic import like so: await import("${path}")`;
            if (isCSSRequest(path)) {
              extra = `It looks like you are resolving styles. If you are adding a link tag, replace with this:

<style global>
@import "${path}";
</style>
`;
            } else if (isScriptRequest(path)) {
              extra = `It looks like you are resolving scripts. If you are adding a script tag, replace with this:

<script type="module" src={(await import("${path}?url")).default}><\/script>

or consider make it a module like so:

<script type="module" hoist>
	import MyModule from "${path}";
<\/script>
`;
            }
            warn(args.logging, `deprecation`, `${bold("Astro.resolve()")} is deprecated. We see that you are trying to resolve ${path}.
${extra}`);
            return "";
          }
          return astroGlobal.resolve(path);
        },
        slots: Object.fromEntries(Object.entries(slots || {}).map(([slotName]) => [slotName, true])),
        privateRenderSlotDoNotUse(slotName) {
          return renderSlot(result, slots ? slots[slotName] : null);
        },
        async privateRenderMarkdownDoNotUse(content, opts) {
          let [mdRender, renderOpts] = markdownRender;
          let parser = null;
          if (Array.isArray(mdRender)) {
            renderOpts = mdRender[1];
            mdRender = mdRender[0];
          }
          if (typeof mdRender === "string") {
            const mod = await import(mdRender);
            parser = mod.default;
          } else if (mdRender instanceof Promise) {
            const mod = await mdRender;
            parser = mod.default;
          } else if (typeof mdRender === "function") {
            parser = mdRender;
          } else {
            throw new Error("No Markdown parser found.");
          }
          const { code } = await parser(content, __spreadValues(__spreadValues({}, renderOpts), opts != null ? opts : {}));
          return code;
        }
      };
    },
    resolve,
    _metadata: {
      renderers,
      pathname,
      experimentalStaticBuild
    }
  };
  return result;
}
export {
  createResult
};
