// astro/dist/vite-plugin-env/index.js
import MagicString from "magic-string";
import { fileURLToPath } from "url";
import { loadEnv } from "vite";
function getPrivateEnv(viteConfig, astroConfig) {
  var _a;
  let envPrefixes = ["PUBLIC_"];
  if (viteConfig.envPrefix) {
    envPrefixes = Array.isArray(viteConfig.envPrefix) ? viteConfig.envPrefix : [viteConfig.envPrefix];
  }
  const fullEnv = loadEnv(viteConfig.mode, (_a = viteConfig.envDir) != null ? _a : fileURLToPath(astroConfig.projectRoot), "");
  const privateKeys = Object.keys(fullEnv).filter((key) => {
    if (typeof process.env[key] !== "undefined")
      return false;
    for (const envPrefix of envPrefixes) {
      if (key.startsWith(envPrefix))
        return false;
    }
    return true;
  });
  if (privateKeys.length === 0) {
    return null;
  }
  return Object.fromEntries(privateKeys.map((key) => [key, fullEnv[key]]));
}
function referencesPrivateKey(source, privateEnv) {
  for (const key of Object.keys(privateEnv)) {
    if (source.includes(key))
      return true;
  }
  return false;
}
function envVitePlugin({ config: astroConfig }) {
  let privateEnv;
  let config;
  return {
    name: "astro:vite-plugin-env",
    enforce: "pre",
    configResolved(resolvedConfig) {
      config = resolvedConfig;
      if (config.envPrefix) {
      }
    },
    async transform(source, id, options) {
      const ssr = (options == null ? void 0 : options.ssr) === true;
      if (!ssr)
        return source;
      if (!source.includes("import.meta"))
        return source;
      if (!/\benv\b/.test(source))
        return source;
      if (typeof privateEnv === "undefined") {
        privateEnv = getPrivateEnv(config, astroConfig);
      }
      if (!privateEnv)
        return source;
      if (!referencesPrivateKey(source, privateEnv))
        return source;
      const s = new MagicString(source);
      s.prepend(`import.meta.env = new Proxy(import.meta.env, {get(target, prop, reciever) {const PRIVATE = ${JSON.stringify(privateEnv)};if (typeof PRIVATE[prop] !== 'undefined') {return PRIVATE[prop];}return Reflect.get(target, prop, reciever);}});
`);
      return s.toString();
    }
  };
}
export {
  envVitePlugin as default
};
